### ISSUE TITLE
MySQL community 5.7 fails due to temporary password

### ISSUE THREAD

**Issue by xenithorb** [geerlingguy/ansible-role-mysql#145]
```
TASK [geerlingguy.mysql : Disallow root login remotely] ************************
failed: [54.84.33.15] (item=DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')) => {"changed": false, "cmd": ["mysql", "-NBe", "DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')"], "delta": "0:00:00.006462", "end": "2016-07-16 22:22:04.457461", "failed": true, "item": "DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')", "rc": 1, "start": "2016-07-16 22:22:04.450999", "stderr": "ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: NO)", "stdout": "", "stdout_lines": [], "warnings": []}
```

Could use something like: 

``` yaml
- name: Acquire temporary root pass
  shell: > 
    awk -F': ' '$0 ~ "temporary password"{print $2}' {{ mysql_log_error }}
  register: mysql_root_temp_password
  when: '5.7.' in mysql_cli_version.stdout

- name: Set temp root pass as root password
  set_fact:
    mysql_root_password: "{{ mysql_root_temp_password.stdout }}"
  when: mysql_root_temp_password is defined
```

**Comment by geerlingguy** [comment]
Hmm... which OS/version are you on, and are you installing from OS repos or from the MySQL repo?

I've been testing with 5.7.x on Ubuntu 16.04 and CentOS 7 and haven't run into this issue (see, for example, Drupal VM: https://github.com/geerlingguy/drupal-vm

**Comment by xenithorb** [comment]
OS: CentOS 7
Repo: MySQL Community 5.7

This is how im calling it:

```yaml
  roles:
    - role: graymonarch.InstallSwapfile
    - role: geerlingguy.sonar
      sonar_version: 5.6
      sonar_download_url: "https://sonarsource.bintray.com/Distribution/sonarqube/sonarqube-{{ sonar_version }}.zip"
      sonar_version_directory: "sonarqube-{{ sonar_version }}"
      # For Dep: geerlingguy.mysql
      mysql_syslog_tag: mysqld
      mysql_daemon: mysqld
      mysql_log_error: /var/log/mysqld.log
      mysql_pid_file: /var/run/mysqld/mysqld.pid
      mysql_packages:
        - mysql-community-client
        - mysql-community-server
        - mysql-community-libs
        # The below are defined in __mysql_packages (defaults) but are overwritten by the above being defined
        - MySQL-python
        - perl-DBD-MySQL
      # For Dep: geerlinguy.java
      java_packages:
        - java-1.8.0-openjdk
```

**Comment by xenithorb** [comment]
I am attempting to work on this here: https://github.com/geerlingguy/ansible-role-mysql/compare/master...graymonarch:mysql57_root_temp_pw?expand=1

But I am running into a consistent problem:

```
TASK [geerlingguy.mysql : ALTER USER to comply with 5.7 password requirements] *
fatal: [54.87.241.240]: FAILED! => {"changed": true, "cmd": "mysql -u root -NBe 'ALTER USER \"root\"@\"localhost\" IDENTIFIED WITH mysql_native_password BY \"CXv5G6uwmhdnYUqharxm\";' #when: (mysql_install_packages | bool) and ('5.7.' in mysql_cli_version.stdout)", "delta": "0:00:00.005622", "end": "2016-07-21 16:30:50.894515", "failed": true, "rc": 1, "start": "2016-07-21 16:30:50.888893", "stderr": "", "stdout": "Please use --connect-expired-password option or invoke mysql in interactive mode.", "stdout_lines": ["Please use --connect-expired-password option or invoke mysql in interactive mode."], "warnings": []}
```
Which either happens with your unmodified role or mine, trying to take advantage of the temporary password....

##### Notes

If I get on a system that has this modified role partially run, I am ably to run `mysql` as root and it does in fact bring up the MySQL console as it's supposed to. This means at least that it's acquiring the temporary password correctly. 

The problem ensues when you attempt to try and do _anything_ after logging in, where you are then presented with:

```
mysql> show databases;
ERROR 1820 (HY000): You must reset your password using ALTER USER statement before executing this statement.
```
I'm not entirely sure how to get around this at the moment, and could sure use some help... As you can see above I made an attempt to add an `ALTER  USER` statement specific for `"root"@"localhost"` but I also can not run that as it seems to be requiring that it be done manually from within the mysql console.

**Comment by xenithorb** [comment]
I have this currently working in the diff here: https://github.com/geerlingguy/ansible-role-mysql/compare/master...graymonarch:mysql57_root_temp_pw?expand=1

The key was in looking for the `Please use --connect-expired-password option or invoke mysql in interactive mode.` error and then actioning on that, because it means a temporary password is in place, or rather technically an expired password. (but in this case on first install it's a temporary one that can be found in the log)

The additional steps to the `secure-installation.yml` playbook in plain english are as follows:

1. Acquire the temporary password made during install-time from `{{ mysql_log_error }}` and register the shell command's standard output as `{{ mysql_root_temp_password }}`
2. Write out the temporary password to `~/.my.cnf` because `mysql` will error out if you attempt to use the password on the commandline
3. Run a simple benign command `SELECT NOW();` in order to determine if we can connect to the mysql engine and register that as `{{ result }}` in order to check for the string: `Please use --connect-expired-password option or invoke mysql in interactive mode.` in the following task:
4. Check for the above string by looking for "connect-expired-password" in `{{ result['stdout'] }}` and run an `ALTER USER` mysql command before all other commands. MySQL 5.7.9+ will not allow us to do any other command UNTIL an `ALTER USER` is run to reset the root password from the expired temporary password
5. Rewrite the `~/.my.cnf` file in order to reflect the new `{{ mysql_root_password }}` 

The reason I did it this way instead reorganizing  the subsequent ALTER USER commands that look very similar is because I didn't want to mess up the flow of the playbook for other versions. The chief reason was also because the command you run to determine all of the root hosts is not able to be run until you change the password with `ALTER USER`

I would love to turn this into a solid PR for you, but I need a guiding hand with regard to when: filters and determining when to run my additions based on 5.7.9+ and maintaining good idempotency.

**Comment by trevorr** [comment]
In case it's useful to anyone else who doesn't need MySQL 5.7, here's a workaround to use MySQL 5.6:

```
- hosts: database_public
  become: yes
  become_user: root
  vars:
    mysql_repo_url: http://dev.mysql.com/get/mysql57-community-release-el6-9.noarch.rpm
    mysql_enablerepo: mysql56-community
  pre_tasks:
    - name: install mysql community repos
      yum:
        name: "{{ mysql_repo_url }}"
        state: present
      register: mysql_repo
      until: '"failed" not in mysql_repo'
      retries: 5
      delay: 10
    - name: disable mysql 5.7 repo
      shell: yum-config-manager --disable mysql57-community
      when: mysql_repo|changed
    - name: enable mysql 5.6 repo
      shell: yum-config-manager --enable mysql56-community
      when: mysql_repo|changed
  roles:
    - geerlingguy.mysql
```

More information: http://dev.mysql.com/doc/refman/5.7/en/linux-installation-yum-repo.html#yum-repo-select-series

**Comment by scutdk** [comment]
also run into this problem. @xenithorb do you have any usable patch

**Comment by pangdudu** [comment]
Same here on: Ubuntu 16.04.1 LTS

`TASK [geerlingguy.mysql : Disallow root login remotely] ************************
failed: [database] (item=DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')) => {"changed": false, "cmd": ["mysql", "-NBe", "DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')"], "delta": "0:00:00.075176", "end": "2016-12-01 13:17:55.975484", "failed": true, "item": "DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')", "rc": 1, "start": "2016-12-01 13:17:55.900308", "stderr": "ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: NO)", "stdout": "", "stdout_lines": [], "warnings": []}`

**Comment by geerlingguy** [comment]
Closing to consolidate discussion in https://github.com/geerlingguy/ansible-role-mysql/issues/57
